worker/lease/store.go

```diff
@@ -0,0 +1,422 @@
+		}
+		return errors.Trace(err)
+	}
+}
+
+// UnpinLease (lease.Store) removes the record indicated by the input
+// key and entity from the lease pin table, indicating that the entity
+// no longer requires the lease to be pinned.
+// When there are no entities associated with a particular lease,
+// it is determined not to be pinned, and can expire normally.
+func (s *Store) UnpinLease(key lease.Key, entity string, stop <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := make(chan error)
+
+	go func() {
+		q := `
+DELETE FROM lease_pin
+WHERE  uuid = (
+    SELECT p.uuid
+    FROM   lease_pin p
+           JOIN lease l ON l.uuid = p.lease_uuid
+           JOIN lease_type t ON l.lease_type_id = t.id
+    WHERE  t.type = ?
+    AND    l.model_uuid = ?
+    AND    l.name = ?
+    AND    p.entity_id = ?   
+)`[1:]
+
+		stmt, err := s.getPrepared(ctx, "UnpinLease", q)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		_, err = stmt.ExecContext(ctx, key.Namespace, key.ModelUUID, key.Lease, entity)
+		errCh <- err
+	}()
+
+	select {
+	case <-stop:
+		cancel()
+		return errors.Trace(<-errCh)
+	case err := <-errCh:
+		cancel()
+		return errors.Trace(err)
+	}
+}
+
+// Pinned (lease.Store) returns all leases that are currently pinned,
+// and the entities requiring such behaviour for them.
+func (s *Store) Pinned() (map[lease.Key][]string, error) {
+	q := `
+SELECT   l.uuid, t.type, l.model_uuid, l.name, p.entity_id
+FROM     lease l 
+         JOIN lease_type t ON l.lease_type_id = t.id
+         JOIN lease_pin p on l.uuid = p.lease_uuid
+ORDER BY l.uuid`[1:]
+
+	stmt, err := s.getPrepared(context.Background(), "Pinned", q)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	rows, err := stmt.Query()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	seen := set.NewStrings()
+	result := make(map[lease.Key][]string)
+	for rows.Next() {
+		var leaseUUID string
+		var key lease.Key
+		var entity string
+
+		if err := rows.Scan(&leaseUUID, &key.Namespace, &key.ModelUUID, &key.Lease, &entity); err != nil {
+			_ = rows.Close()
+			return nil, errors.Trace(err)
+		}
+
+		if !seen.Contains(leaseUUID) {
+			result[key] = []string{entity}
+			seen.Add(leaseUUID)
+		} else {
+			result[key] = append(result[key], entity)
+		}
+	}
+
+	return result, errors.Trace(rows.Err())
+}
+
+// getPrepared returns a prepared statement for the input name,
+// ensuring that the first call for a given name caches the statement.
+// thereafter the statement is returned from the cache.
+func (s *Store) getPrepared(ctx context.Context, name string, stmt string) (*sql.Stmt, error) {
+	s.cacheMu.RLock()
+	if cachedStmt, ok := s.cache[name]; ok {
+		s.cacheMu.RUnlock()
+		return cachedStmt, nil
+	}
+	s.cacheMu.RUnlock()
+
+	s.cacheMu.Lock()
+	defer s.cacheMu.Unlock()
+
+	prepared, err := s.db.PrepareContext(ctx, stmt)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	s.cache[name] = prepared
+	return prepared, nil
```

Comment:
I'm unsure if this will work, I'll investigate, but if you have a leadership change do we keep the prepared statement for all followers?